### PR TITLE
fix doubly shell quoting args to `spack spec`

### PR DESCRIPTION
**Pull request by cosmicexplorer**
### Problem
The following command fails:
```bash
$ spack spec 'emacs toolkit=gtk ^ gcc languages=jit' 
==> Error: invalid values for variant "toolkit" in package "emacs": ['gtk ^ gcc languages=jit']
```

But this appears to be a valid spec when parsed directly:
```bash
$ spack python
Spack version 0.17.1
Python 3.10.2, Linux x86_64
>>> from spack.spec import Spec
>>> print(Spec('emacs toolkit=gtk ^ gcc languages=jit').to_yaml())
spec:
  _meta:
    version: 2
  nodes:
  - name: emacs
    versions:
    - ':'
    parameters:
      toolkit:
      - gtk
    concrete: false
    dependencies:
    - name: gcc
      hash: ku4hvzuxvhxqv6b7qonr2ic4lsxsilxt
      type: []
    hash: n3yengnkkhcyuhjw2sehq6u7tnushgau
  - name: gcc
    versions:
    - ':'
    parameters:
      languages:
      - jit
    concrete: false
    hash: ku4hvzuxvhxqv6b7qonr2ic4lsxsilxt
```

### Solution
- Avoid shell quoting arguments to `spack spec` multiple times before joining them.

### Result
```bash
$ spack spec 'emacs toolkit=gtk ^ gcc languages=jit'
Input spec
--------------------------------
emacs toolkit=gtk
    ^gcc languages=jit

Concretized
--------------------------------
emacs@master%gcc@10.3.1~X+native~tls toolkit=gtk arch=linux-alpine3-zen3
    ^autoconf@2.71%gcc@10.3.1 arch=linux-alpine3-zen3
    ^automake@1.16.5%gcc@10.3.1 arch=linux-alpine3-zen3
    ^gcc@11.2.0%gcc@10.3.1~binutils+bootstrap~graphite~nvptx~piclibs+strip languages=jit arch=linux-alpine3-zen3
... (truncated)
```

### CODE REVIEW

**Comment by haampie** [comment]
Can you check the failing tests @cosmicexplorer? To me this change looks sensible, we shouldn't rely on the shell splitting arguments, but just pass a string with whitespaces to the parser.

File: lib/spack/spack/test/cmd/spec.py [hunk 1]
```diff
@@ -55,6 +55,16 @@ def test_spec_concretizer_args(mutable_config, mutable_database):
...
 
+def test_spec_parse_arg_quoting():
+    """Assert that certain spec constructions work when provided as arguments."""
+    # Verify that we can provide multiple key=value variants to multiple separate
+    # packages within a spec string.
+    output = spec('multivalue-variant fee=barbaz ^ a foobar=baz')
+
+    assert 'fee=barbaz' in output
+    assert 'foobar=baz' in output
+
```
Comment (becker33) [inline lib/spack/spack/test/cmd/spec.py]: I believe that this test succeeds even without the changes in this PR, can you double-check on that?

**Review by becker33 — COMMENTED** [review]
@trws I think this addresses your question we discussed recently

File: lib/spack/spack/test/cmd/spec.py (hunk 1 above)
Comment (trws) [reply lib/spack/spack/test/cmd/spec.py]: You have truly fantastic timing @cosmicexplorer.  Would you mind testing: 
- `spack spec ‘gcc@4.9.3 cflags=“-Os -pipe” cxxflags=“-Os -pipe”`
- `spack spec ‘gcc@4.9.3 cflags=-Os -pipe cxxflags=-Os -pipe`
and see what comes out?

This should definitely help, but I think we may still have an over-eager glob on options with a VAL token.

**Comment by trws** [comment]
Absolutely fabulous: 
```
~/spack fix-dep-kv-flags *1 ?3                                                                                                                                  ✘ 2 scogland1@chimera
❯ spack spec 'gh cflags="-Os -pipe" cxxflags="-Os -pipe"'
Input spec
--------------------------------
gh cflags="-Os -pipe" cxxflags="-Os -pipe"
```

This would be a good one to include as a test, because it was most definitely broken before this.

**Comment by trws** [comment]
> Can you check the failing tests @cosmicexplorer? To me this change looks sensible, we shouldn't rely on the shell splitting arguments, but just pass a string with whitespaces to the parser.

The test I see breaking looks very much like a broken test to me.  It relies on `meh cflags=-u -g` parsing as `meh cflags=“-u -g”`, which would be somewhat like how Dockerfiles parse the ENV attribute, but that’s a *much* trickier parse, with a whole lot more corner cases in it.

File: lib/spack/spack/test/cmd/spec.py (hunk 1 above)
Comment (cosmicexplorer) [reply lib/spack/spack/test/cmd/spec.py]: @becker33:
> I believe that this test succeeds even without the changes in this PR, can you double-check on that?

I orchestrated the commits so that the test is added before the fix, and this test indeed fails without the fix:
```
spack.variant.InvalidVariantValueError: invalid values for variant "fee" in package "multivalue-variant": ['barbaz ^ a foobar=baz']
```

@trws:
```bash
> spack spec 'gcc@4.9.3 cflags="-Os -pipe" cxxflags="-Os -pipe"'
...
gcc@4.9.3%gcc@11.2.1 cflags="-Os -pipe" cxxflags="-Os -pipe" ~binutils+bootstrap~graphite~nvptx~piclibs~strip languages=c,c++,fortran patches=0689c26,27e9829,7112abb,f5643d9,f73de87 arch=linux-alpine3-zen3
...
> spack spec 'gcc@4.9.3 cflags=-Os -pipe cxxflags=-Os -pipe'
==> Error: Cannot specify variant "pipe" twice
```

Regarding the second command line with `cflags=-Os -pipe`: that one is currently accepted by spack on `develop`, so while I think this change should still be made, it may break some users. How should we notify users about this?

Additionally, we can verify the issue from https://github.com/spack/spack/pull/29282#issuecomment-1057496052 is fixed with the `--yaml` output:

```
> spack spec --yaml 'gh cflags="-Os -pipe" cxxflags="-Os -pipe"' | yq
...
        cflags:
          - -Os
          - -pipe
...
> git checkout develop
> spack spec --yaml 'gh cflags="-Os -pipe" cxxflags="-Os -pipe"' | yq
...
        cflags:
          - '"-Os'
          - -pipe" cxxflags="-Os
          - -pipe"
...
```

So it looks like that would absolutely be a good one to include as a new test.

**Comment by cosmicexplorer** [comment]
@trws:
> The test I see breaking looks very much like a broken test to me. It relies on `meh cflags=-u -g` parsing as `meh cflags="-u -g"`, which would be somewhat like how Dockerfiles parse the ENV attribute, but that’s a much trickier parse, with a whole lot more corner cases in it.

I fixed `test_parse_spec_flags_with_spaces()` just now, but this implies another decision will have to be made:
> Regarding the second command line with `cflags=-Os -pipe`: that one is currently accepted by spack on develop, so while I think this change should still be made, it may break some users. How should we notify users about this?

Since this change is actually necessary in order to e.g. write `spack spec 'emacs toolkit=gtk ^ gcc languages=jit'`, I agree that we emphatically *should* make this breaking change in order to enable that type of command line to be written at all (while command lines relying on `cflags=-Os -pipe` to work without quotes can just add quotes). 

EDIT: I just realized we should be able to actually shlex `spack spec` cli args and look for cases where a `-*` argument follows a `*flags=...` argument, and print out a warning if detected. I'll look into doing that so we can hopefully avoid making anyone too confused.

**Comment by cosmicexplorer** [comment]
Ok, just generated a really nice error report, which also has a test added for it in af00a25. It looks like:

```
> spack spec 'gcc@4.9.3 cflags=-Os -pipe'     
==> Error: trying to set variant "pipe" in package "gcc", but the package has no such variant [happened during concretization of gcc@4.9.3 cflags="-Os" ~pipe]

Some compiler or linker flags were provided without quoting their arguments,
which now causes spack to try to parse the *next* argument as a spec component
such as a variant instead of an additional compiler or linker flag. If the
intent was to set multiple flags, try quoting them together as described below.

Possible flag quotation errors (with the correctly-quoted version after the =>):
(1) cflags=-Os -pipe => cflags="-Os -pipe"
```

For multiple such cases (probably unlikely, but necessary in order to ensure that the arguments causing the error are represented in the error report):
```
> spack spec 'gcc@4.9.3 cflags=-Os -pipe cflags=-I /usr/include'
==> Error: No installed spec matches the hash: 'usr'

Some compiler or linker flags were provided without quoting their arguments,
which now causes spack to try to parse the *next* argument as a spec component
such as a variant instead of an additional compiler or linker flag. If the
intent was to set multiple flags, try quoting them together as described below.

Possible flag quotation errors (with the correctly-quoted version after the =>):
(1) cflags=-Os -pipe => cflags="-Os -pipe"
(2) cflags=-I /usr/include => cflags="-I /usr/include"
```

**Comment by trws** [comment]
That looks really good to me.

**Comment by cosmicexplorer** [comment]
gitlab ci had a spurious failure to start the docker daemon and said to try again (https://gitlab.spack.io/spack/spack/-/jobs/2481179), so just force pushed to kick it off again.

**Comment by trws** [comment]
Looks like some github action unit test issues also?

**Comment by cosmicexplorer** [comment]
Ok, the unit test failure on 3.5 appears to be because subscripting `re.Match` objects was introduced in 3.6, so that was switched to use `.group(n)` instead. The coverage changes reported by codecov don't correspond to any code I've actually changed, and coverage has only increased for every file in the "diff" section, so I'm inclined to wait for it to re-run and see if it still complains.

**Comment by cosmicexplorer** [comment]
codecov/project failed again, so I pulled `develop` and tried again, and it failed but with an even smaller reported decrease in coverage, so I just added a test for something that wasn't covered and hoping that allows this to finally pass. Incredibly frustrated with the complete lack of indication as to why a diff produced a change in coverage.

**Comment by trws** [comment]
If this set passes with *only* the codecov failure, we can merge.  Codecov is no longer set as required because of spurious failure issues.

**Comment by cosmicexplorer** [comment]
Looks like it passed!

**Review by trws — DISMISSED** [review]

File: lib/spack/spack/cmd/__init__.py [hunk 2]
```diff
@@ -157,15 +212,28 @@ def parse_specs(args, **kwargs):
...
     if not isinstance(args, six.string_types):
-        sargs = ' '.join(spack.util.string.quote(args))
-    specs = spack.spec.parse(sargs)
-    for spec in specs:
-        if concretize:
-            spec.concretize(tests=tests)  # implies normalize
-        elif normalize:
-            spec.normalize(tests=tests)
+        sargs = ' '.join(args)
+    unquoted_flags = _UnquotedFlags.extract(sargs)
```
Comment (tgamblin) [inline lib/spack/spack/cmd/__init__.py:216]: I'd move this inside the `except` block (just before `if unquoted_flags:`), as it's not needed except when there's an error.

**Review by tgamblin — APPROVED** [review]
